### PR TITLE
Add D3D GPU Sensors

### PR DESCRIPTION
**Pull request by kolos**

### CODE REVIEW

**Comment by PhyxionNL** [comment]
Conflits + formatting.

**Comment by kolos** [comment]
Should work work with NVIDIA too.
Needs someone with a GPU to match up the card with the Windows device identifier.

**Comment by sebastian-dev** [comment]
Geforce 940M,
D3DDisplayDevice.GetDisplayDeviceNames()
`\\?\PCI#VEN_8086&DEV_1916&SUBSYS_1B1D1043&REV_07#3&11583659&0&10#{1ca05180-a699-450a-9a0c-de4fbe3ddd89}`

**Comment by PhyxionNL** [comment]
Anyone know of a method to get PNP identifier from NVIDIA device?

File: LibreHardwareMonitorLib/Interop/Gdi32.cs [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+// ReSharper disable IdentifierTypo
+// ReSharper disable InconsistentNaming
+
+using System.Runtime.InteropServices;
+
+namespace LibreHardwareMonitor.Interop
+{
+    internal static class Gdi32
+    {
+        internal const string DllName = "CfgMgr32.dll";
```
Comment (kolos) [inline LibreHardwareMonitorLib/Interop/Gdi32.cs:15]: this should be Gdi32.dll :)

File: LibreHardwareMonitorLib/Interop/D3dkmdt.cs [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+// ReSharper disable IdentifierTypo
+// ReSharper disable InconsistentNaming
+
+using System.Runtime.InteropServices;
+
+namespace LibreHardwareMonitor.Interop
+{
+    internal static class D3dkmdt
+    {
+        internal enum DXGK_ENGINE_TYPE
```
Comment (kolos) [inline LibreHardwareMonitorLib/Interop/D3dkmdt.cs:15]: Can we just keep the comments for the engine types? I think those were helpful.

Maybe add as a hover tooltip for the sensor?

File: LibreHardwareMonitorLib/Hardware/D3DDisplayDevice.cs [hunk 3]
```diff
@@ -1067,34 +52,34 @@ public static bool GetDeviceInfoByName(string displayDeviceName, out D3DDeviceIn
...
 
-                deviceInfo.Nodes[nodeId] = new D3DDeviceNodeInfo()
+
+                deviceInfo.Nodes[nodeId] = new D3DDeviceNodeInfo
                 {
-                    Id = nodeId,
-                    Name = GetNodeEngineTypeString(nodeMetaData),
-                    RunningTime = nodeInformation.GlobalInformation.RunningTime.QuadPart,
-                    QueryTime = DateTime.Now
+                    Id = nodeId, Name = GetNodeEngineTypeString(nodeMetaData), RunningTime = nodeInformation.GlobalInformation.RunningTime.QuadPart, QueryTime = DateTime.Now
```
Comment (kolos) [inline LibreHardwareMonitorLib/Hardware/D3DDisplayDevice.cs:67]: I think it was more readable with the newlines

Comment (PhyxionNL) [reply LibreHardwareMonitorLib/Hardware/D3DDisplayDevice.cs:67]: Agreed, somehow this got moved to a single line.

**Comment by PhyxionNL** [comment]
I found a way to get these to work on NVIDIA cards as well. There's no proper way to get the PNP id for these as far as I could find, but nvlddmkm seems to write the id in registry. Otherwise it tries to fallback to bus id in registry.

**Comment by Rem0o** [comment]
I think this PR broke AMD GPU detection/sensor detection.

**Comment by PhyxionNL** [comment]
> 
> 
> I think this PR broke AMD GPU detection/sensor detection.

Can you take a look at a fix? 😊

**Comment by Rem0o** [comment]
@PhyxionNL I don't have an AMD GPU, but I got several people reporting that their AMD GPU went missing since this commit.  Just reporting the issue :/

**Comment by PhyxionNL** [comment]
> 
> 
> @PhyxionNL I don't have an AMD GPU to test sadly, but I got several people reporting that their AMD GPU went missing since this commit. Just reporting the issue :/

Ah, same problem. @kolos can you take a look?

**Comment by PhyxionNL** [comment]
https://github.com/Rem0o/FanControl.Releases/issues/256

**Comment by kolos** [comment]
Somehow theres a fixed ```1``` value instead of ```_adapterIndex``` in here:
https://github.com/LibreHardwareMonitor/LibreHardwareMonitor/blob/8717191fc21f46cfdebfde1077331fcbfe021285/LibreHardwareMonitorLib/Hardware/Gpu/AmdGpu.cs#L140

That could be the root cause :)

**Comment by PhyxionNL** [comment]
I expected the same, I have that changed already here
https://github.com/LibreHardwareMonitor/LibreHardwareMonitor/tree/amd
But he reports it's still not working.
https://github.com/Rem0o/FanControl.Releases/issues/256#issuecomment-857754482

**Pull request merged**